File: backend/src/main/java/com/yourapp/app/services/CategoriaService.java
package com.yourapp.app.services;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.yourapp.app.exceptions.ConflictException;
import com.yourapp.app.exceptions.NotFoundException;
import com.yourapp.app.mappers.CategoriaMapper;
import com.yourapp.app.models.dto.CategoriaDto;
import com.yourapp.app.models.entities.Categoria;
import com.yourapp.app.repositories.CategoriaRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class CategoriaService {
    private final CategoriaRepository categoriaRepository;

    // ============================ CREAR CATEGORIA ============================
    @Transactional
    public Categoria crearCategoria(CategoriaDto categoriaDto) {
        if (categoriaRepository.existsByDescripcionAndFueEliminadoFalse(categoriaDto.getDescripcion())) throw new ConflictException("La descripción de la categoria ya está en uso");
        Categoria categoria = CategoriaMapper.toEntity(categoriaDto);
        return categoriaRepository.save(categoria);
    }

    // ============================ ELIMINAR UNA CATEGORIA + SUBCATEGORIAS ============================
    public void eliminarCategoria(Long id) {
        Categoria categoria = obtenerCategoria(id);
        categoria.softDelete();
        categoriaRepository.save(categoria);
    }

    // ============================ OBTENER CATEGORIA ============================
    public Categoria obtenerCategoria(Long categoriaId) {
        Categoria categoria = categoriaRepository.findById(categoriaId).orElseThrow(() -> new NotFoundException("Categoria no encontrada"));
        if (categoria.getFueEliminado()) throw new NotFoundException("Categoria eliminada");
        return categoria;
    }

    // ============================ OBTENER TODAS LAS CATEGORIAS ============================
    public List<Categoria> obtenerTodasLasCategorias() {
        return categoriaRepository.findByFueEliminadoFalse();
    }
}

File: frontend/src/services/categoriaService.js
import { api } from "../api/api.js";
import { ENDPOINTS } from "../api/endpoints.js";

const CategoriaService = {
  // =======================
  // POST /categorias (Crear categoría)
  // =======================
  crearCategoria: async (payload) => {
    const { data } = await api.post(ENDPOINTS.CATEGORIAS.BASE, payload);
    return data;
  },

  // =======================
  // GET /categorias (Obtener todas)
  // =======================
  getCategorias: async () => {
    const { data } = await api.get(ENDPOINTS.CATEGORIAS.BASE);
    return data;
  },

  // =======================
  // PATCH /categorias/{id}/eliminacion
  // =======================
  eliminarCategoria: async (id) => {
    await api.patch(ENDPOINTS.CATEGORIAS.ELIMINAR(id));
  },
};

export default CategoriaService;
